File: app/src/main/java/com/michaeljordanr/memesounds/MainActivity.java
package com.michaeljordanr.memesounds;

import android.content.Intent;
import android.net.Uri;
import android.os.Bundle;
import android.support.v7.app.AppCompatActivity;
import android.support.v7.widget.GridLayoutManager;
import android.support.v7.widget.RecyclerView;
import android.support.v7.widget.SearchView;
import android.support.v7.widget.Toolbar;
import android.view.Menu;
import android.view.MenuItem;
import android.view.View;

import com.github.piasy.rxandroidaudio.PlayConfig;
import com.github.piasy.rxandroidaudio.RxAudioPlayer;
import com.google.gson.Gson;
import com.google.gson.reflect.TypeToken;

import java.lang.reflect.Type;
import java.util.List;

import butterknife.BindView;
import butterknife.ButterKnife;
import io.reactivex.Observer;
import io.reactivex.disposables.Disposable;
import io.reactivex.schedulers.Schedulers;


public class MainActivity extends AppCompatActivity implements AudioAdapter.RecyclerAdapterOnClickListener,
        AudioAdapter.RecyclerAdapterOnLongListener{

    private static final String uriAssetsProvider = "content://" + BuildConfig.APPLICATION_ID + "/";
    private static final String JSON_CONFIG_PATH = "config.json";

    private RxAudioPlayer rxAudioPlayer;
    private AudioAdapter adapter;

    @BindView(R.id.rv_buttons)
    RecyclerView recyclerView;
    @BindView(R.id.toolbar)
    Toolbar toolbar;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);
        ButterKnife.bind(this);
        setSupportActionBar(toolbar);
        if (getSupportActionBar() != null) {
            getSupportActionBar().setTitle(R.string.app_name);
        }

        String jsonSource = Utils.loadJSONFromAsset(this,JSON_CONFIG_PATH);
        Type type = new TypeToken<List<Audio>>() {
        }.getType();
        List<Audio> audioList = new Gson().fromJson(jsonSource, type);

        adapter = new AudioAdapter(this, this, this);
        adapter.setData(audioList);

        GridLayoutManager layoutManager = new GridLayoutManager(this,3);
        recyclerView.setLayoutManager(layoutManager);
        recyclerView.setHasFixedSize(true);
        recyclerView.setAdapter(adapter);

        rxAudioPlayer = RxAudioPlayer.getInstance();
    }

    @Override
    public boolean onCreateOptionsMenu(Menu menu) {
        getMenuInflater().inflate(R.menu.menu_main, menu);

        MenuItem searchItem = menu.findItem(R.id.search);

        SearchView searchView = (SearchView) searchItem.getActionView();
        searchView.setFocusable(false);
        searchView.setQueryHint(getText(R.string.search_hint));
        searchView.setOnQueryTextListener(new SearchView.OnQueryTextListener() {
            @Override
            public boolean onQueryTextSubmit(String query) {
                return false;
            }

            @Override
            public boolean onQueryTextChange(String query) {
                adapter.getFilter().filter(query);
                if(query.equals("")){
                    searchView.setIconified(true);
                }
                return false;
            }
        });
        return true;
    }

    @Override
    public void onClick(Audio audio) {
        play(audio.getAudioName());
    }

    @Override
    public void onLongClick(Audio audio) {
        shareAudio(audio);
    }

    private void play(String audioName){
        PlayConfig audioLoaded = PlayConfig.file(Utils.getFileAudio(this, audioName))
                //PlayConfig.res(getApplicationContext(), resourceId) // or play a raw resource
                .looping(false) // loop or not
                .leftVolume(1.0F) // left volume
                .rightVolume(1.0F) // right volume
                .build(); // build this config and play!

        rxAudioPlayer.play(audioLoaded)
                .subscribeOn(Schedulers.io())
                .subscribe(new Observer<Boolean>() {
                    @Override
                    public void onSubscribe(final Disposable disposable) { }

                    @Override
                    public void onNext(final Boolean aBoolean) { }

                    @Override
                    public void onError(final Throwable throwable) { }

                    @Override
                    public void onComplete() { }
                });
    }

    private void shareAudio(Audio audio){
        Uri uri = Uri.parse(uriAssetsProvider + audio.getAudioName() + Utils.AUDIO_FORMAT);
        Intent intent = new Intent(Intent.ACTION_SEND);
        intent.setType("audio/*");
        intent.putExtra(Intent.EXTRA_STREAM, uri);
        startActivity(Intent.createChooser(
                intent,
                getResources().getString(R.string.share_audio)
                + " \"" + audio.getAudioDescription() + "\"")
        );
    }

}

File: app/src/main/java/com/michaeljordanr/memesounds/MainActivity.kt
package com.michaeljordanr.memesounds

import android.content.Intent
import android.os.Bundle
import android.view.Menu
import android.widget.Toast
import androidx.appcompat.app.AppCompatActivity
import androidx.appcompat.widget.SearchView
import androidx.appcompat.widget.Toolbar
import androidx.viewpager2.widget.ViewPager2
import com.google.android.material.tabs.TabLayout
import com.google.android.material.tabs.TabLayoutMediator
import com.google.android.play.core.appupdate.AppUpdateInfo
import com.google.android.play.core.appupdate.AppUpdateManager
import com.google.android.play.core.appupdate.AppUpdateManagerFactory
import com.google.android.play.core.install.model.AppUpdateType
import com.google.android.play.core.install.model.UpdateAvailability
import com.google.android.play.core.tasks.Task


class MainActivity : AppCompatActivity(), AudioListFragment.AudioListFragmentListener {

    companion object {
        const val JSON_CONFIG_PATH = "config.json"
        const val APP_UPDATE_REQUEST_CODE = 1137
    }

    var selectedTab = AudioListFragment.AudioListType.ALL

    lateinit var viewPager: ViewPager2
    lateinit var tabs: TabLayout


    private val viewPagerAdapter by lazy {
        ViewPagerAdapter(this)
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_main)
        val toolbar = findViewById<Toolbar>(R.id.toolbar)
        setSupportActionBar(toolbar)
        if (supportActionBar != null) {
            supportActionBar!!.setTitle(R.string.app_name)
        }

        viewPager = findViewById(R.id.view_pager)
        tabs = findViewById(R.id.tab_layout)

        viewPager.adapter = viewPagerAdapter

        TabLayoutMediator(tabs, viewPager,
                TabLayoutMediator.TabConfigurationStrategy { tab, position ->
                    when (position) {
                        0 -> {
                            tab.text = getString(R.string.all)
                        }
                        1 -> {
                            tab.text = getString(R.string.bookmarks)
                        }
                    }
                }).attach()

        viewPager.registerOnPageChangeCallback(object : ViewPager2.OnPageChangeCallback() {
            override fun onPageSelected(position: Int) {
                super.onPageSelected(position)
                selectedTab = if (position == AudioListFragment.AudioListType.FAVORITES.value) {
                    AudioListFragment.AudioListType.FAVORITES
                } else {
                    AudioListFragment.AudioListType.ALL
                }
            }
        })

        checkUpdate()
    }

    override fun onCreateOptionsMenu(menu: Menu): Boolean {
        menuInflater.inflate(R.menu.menu_main, menu)

        val searchItem = menu.findItem(R.id.search)

        val searchView = searchItem.actionView as SearchView
        searchView.isFocusable = false
        searchView.queryHint = getText(R.string.search_hint)
        searchView.setOnQueryTextListener(object : SearchView.OnQueryTextListener {
            override fun onQueryTextSubmit(query: String): Boolean {
                return false
            }

            override fun onQueryTextChange(query: String): Boolean {
                viewPagerAdapter.setQuery(query)
                viewPager.adapter = viewPagerAdapter
                goToPage(selectedTab)

                val params = HashMap<String, String>()
                params["query"] = query

                return false
            }
        })
        return true
    }

    override fun goToPage(audioListType: AudioListFragment.AudioListType) {
        if (audioListType == AudioListFragment.AudioListType.FAVORITES) {
            viewPager.currentItem = 1
        } else {
            viewPager.currentItem = 0
        }
        viewPager.adapter = viewPagerAdapter
    }

    private fun checkUpdate() {
        // Creates instance of the manager.
        val appUpdateManager = AppUpdateManagerFactory.create(baseContext)

        // Returns an intent object that you use to check for an update.
        val appUpdateInfoTask = appUpdateManager.appUpdateInfo

        // Checks that the platform will allow the specified type of update.
        appUpdateInfoTask.addOnSuccessListener { appUpdateInfo ->
            if (appUpdateInfo.updateAvailability() == UpdateAvailability.UPDATE_AVAILABLE
                    // For a flexible update, use AppUpdateType.FLEXIBLE
                    && appUpdateInfo.isUpdateTypeAllowed(AppUpdateType.FLEXIBLE)
            ) {
                // Request the update.
                appUpdateManager.startUpdateFlowForResult(
                        // Pass the intent that is returned by 'getAppUpdateInfo()'.
                        appUpdateInfo,
                        // Or 'AppUpdateType.FLEXIBLE' for flexible updates.
                        AppUpdateType.FLEXIBLE,
                        // The current activity making the update request.
                        this,
                        // Include a request code to later monitor this update request.
                        APP_UPDATE_REQUEST_CODE)
            }
        }
    }

    override fun onActivityResult(requestCode: Int, resultCode: Int, data: Intent?) {
        super.onActivityResult(requestCode, resultCode, data)

        if (requestCode == APP_UPDATE_REQUEST_CODE) {
            if (resultCode != RESULT_OK) {
                Toast.makeText(baseContext, "Por que você fez isso, sua lambisgóia?!", Toast.LENGTH_LONG).show()
            }
        }
    }
}
